Storage Add-Ons: Add tracks event

jeyip opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->
Related to https://github.com/Automattic/martech/issues/2027
Tracking Issue https://github.com/Automattic/martech/issues/2023

## Proposed Changes

* Adds user interaction tracks event to the storage add-on dropdown
* Passes tracks event callback through plans features main component to avoid creating a dependency to the Calypso environment for the [2023 plan features component](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plan-features-2023-grid/index.tsx)

## Note
I will formally register this tracks event once this PR is approved

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Check out this PR or use Calypso Live
* Smoke test the onboarding signup flow with business or ecommerce plans `/start?flags=plans%2Fupgradeable-storage`
* Verify that, when interacting with the dropdown, a tracks event is dispatched using either the tracks vigilante chrome dev tool or or searching for `.gif storage` in the dev tools network tab

<img width="844" alt="Screenshot 2023-09-17 at 3 44 13 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/eae6591a-1a47-4084-a4dd-b3e6b03c8538">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

jeyip committed:
Add tracks event to storage add-on dropdown

jeyip commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -700,6 +700,9 @@ const PlansFeaturesMain = ( {
 							planTypeSelectorProps={ planTypeSelectorProps }
 							ref={ plansComparisonGridRef }
+							onStorageAddOnClick={ () => {
+								recordTracksEvent( 'calypso_plans_page_storage_add_on_click' );
```
**Note:**

I gave some thought to other, relevant information we wanted to track with this event. Given, however, that we're limiting the feature to the signup onboarding flow ( so we already know the flowname and the signup flow ), I think simply tracking the event is enough. Definitely open to suggestions though

chriskmnds commented on client/my-sites/plan-features-2023-grid/components/storage-add-on-dropdown.tsx:
```diff
@@ -92,9 +94,10 @@ export const StorageAddOnDropdown = ( {
-			}
+			onChange={ ( { selectedItem }: { selectedItem: { key: WPComStorageAddOnSlug } } ) => {
+				onStorageAddOnClick && onStorageAddOnClick();
+				setSelectedStorageOptionForPlan( { addOnSlug: selectedItem?.key || '', planSlug } );
```
Under what conditions would an empty `''` `addOnSlug` make sense for this call?

jeyip replied:
You're right 👍 I've updated the logic such that the data store storage option action is only dispatched when there _is_ a meaningful key. Thank you! https://github.com/Automattic/wp-calypso/pull/81816/commits/6aefa8524f8c74402b4254c1e585fad5c19a497f

_Edit:_

Odd -- the changes didn't show up for a little while. Not sure if there's something wonky with Github, but if the changes don't show up on your end, try looking at the updates in a different browser

jeyip committed:
Only dispatch setSelectedStorageOptionForPlan with valid key

jdc91 commented:
LGTM 🚢 

That is some next level prop drilling 😓 . Have you thought about whether we should probably introduce a `trackingCallbacks` optional context prop to avoid this?

CC - @chriskmnds

jeyip commented:
>  Have you thought about whether we should probably introduce a trackingCallbacks optional context prop to avoid this?

I agree that the prop drilling is unpleasant. I really like this idea -- @chriskmnds how does that sound to you? I'm unsure if it belongs in `PlansGridContextProvider` though 🤔

chriskmnds commented:
> Have you thought about whether we should probably introduce a trackingCallbacks optional context prop to avoid this?

> I agree that the prop drilling is unpleasant. I really like this idea -- @chriskmnds wondering if you have opinions on this and using the PlansGridContextProvider ( I'm unsure if it fits in there )

I like the idea of bringing these together too. I think there is one more case of tracking that needs to be refactored as part of the NPM Refactor (tracked in https://github.com/Automattic/wp-calypso/issues/79006). Might be worth exploring this abstraction/grouping there. :-)

southp commented:
It would be really nice to consolidate it so that we can both decoupling the tracks event lib dependency and avoiding prop drilling. Before we have that, it looks like we need to stay with this prop drilling for a bit longer :) 

@jeyip I do have some questions regarding the event itself though. Could you help me clarify what we want to analyze from this event? e.g. 

* In its current form, it can only track the number of clicks, not which storage add-on is clicked. Are we sure we don't want to know that?
* Do we need to track which storage add-on is eventually picked?

jeyip committed:
Support multiple checkout cart items in buildUpgradeFunction

southp reviewed (changes requested):
Thanks for clarifying your thought :) 

> If we were analyzing storage add-ons, I assumed that there'd be a funnel that we could create with a checkout event which would have the relevant information ( about what exactly was purchased in check out ). If you're aware that that's not the case, then I can definitely track the storage add-on that's selected here as well.

We can create a funnel like this indeed. However, from there we'd want to do more granular analysis, e.g.

1. The actual distribution of which item is clicked.
2. The distribution of which item is eventually selected.
3. Given a selected add-on, what is the eventual conversion rate? 

To enable these forseeable granular analysis, let's add:

1. Which item is clicked to the click event.
2. A new event tracking which item is eventually selected.

jeyip committed:
DRY getPlanCartItem

jeyip committed:
Include add on slug in storage dropdown option click event

jeyip committed:
Add tracks event for storage add on upgrade click

jeyip commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -439,6 +441,7 @@ export class PlanFeatures2023Grid extends Component< PlanFeatures2023GridType >
 			product_slug: storageAddOn.productSlug,
 			quantity: storageAddOn.quantity,
 			volume: 1,
+			extra: { feature_slug: selectedStorageOption },
```
**Note:**
Because the add on `product_slug` is identical for both 50GB and 100GB storage add-ons, we add `feature_slug` data to Add On Cart Item extra parameters. That way we can dispatch a tracks event with a unique, human readable slug in https://github.com/Automattic/wp-calypso/pull/81816/files#diff-18214fd84316736ff6662f03fa201b6255f4c20f277c0eb683b5e36e9c9dea6dR300-R317

This looks reasonable to me, but if it is improper usage of the `extra` property, I'm happy to switch things up.

More context about why we don't simply retrieve the selected storage option feature slug from the data store in `plans-features-main` can be found in https://github.com/Automattic/wp-calypso/pull/81340#discussion_r1331477865.

sirbrillig replied:
`extra` has become kind of a dumping ground for all sorts of metadata that the client needs to pass to the server when adding a product, so I think this is probably fine.

southp commented:
@jeyip I'm moving this back to In Progress since it needs to be rebased now. Please feel free to move it back to Needs Review once it's done :)

sirbrillig commented on client/state/sites/hooks/use-billing-history.ts:
```diff
@@ -13,7 +13,7 @@ const fetchPastBillingTransactions = () =>
 	} );
 
-export const usePastBillingTransactions = () => {
+export const usePastBillingTransactions = ( isInSignup: boolean ) => {
```
If the only purpose of `isInSignup` is to disable this hook, I think that a disabled parameter would be more semantic and reusable.

```suggestion
export const usePastBillingTransactions = ( disabled: boolean ) => {
```

jeyip replied:
This is a great suggestion. Although the changes actually end up being unrelated to this PR, I'm making the changes since they're a small update 😃

_Edit:_ Resolved in [e8adbc6](https://github.com/Automattic/wp-calypso/pull/81816/commits/e8adbc618cbd4d717eff87fd5684fb6744047159)

sirbrillig commented on client/state/sites/hooks/use-billing-history.ts:
```diff
@@ -25,5 +25,6 @@ export const usePastBillingTransactions = () => {
 		billingTransactions: data?.billing_history || null,
 		isLoading,
```
When a `useQuery` hook is disabled, its `isLoading` will be `true` forever (see p7H4VZ-4nx-p2), which is not always expected and can cause pages to hang for consumers, so it might be a good idea to make this false if disabled.

```ts
isLoading: disabled ? false : isLoading,
```

jeyip replied:
TIL -- thanks for the heads up 👍 Making these recommended changes as well

_Edit_: Resolved in [2efdcdd](https://github.com/Automattic/wp-calypso/pull/81816/commits/2efdcdd3d34240a27144eea10bb356a8f786e027)

jeyip committed:
Update useBillingHistory isInSignup param to disabled

jeyip committed:
Set isLoading to false if query is disabled

When a useQuery hook is disabled, its isLoading will be true forever (see p7H4VZ-4nx-p2), which is not always expected and can cause pages to hang for consumers, so it might be a good idea to make this false if disabled.

jeyip commented on client/state/sites/hooks/use-billing-history.ts:
```diff
@@ -13,7 +13,7 @@ const fetchPastBillingTransactions = () =>
 	} );
 
-export const usePastBillingTransactions = ( isInSignup: boolean ) => {
+export const usePastBillingTransactions = ( disabled: boolean ) => {
```
**Note:**

Changes in this file are unrelated to tracks events, but they're small, relatively safe, and meaningful updates as suggested by Payton https://github.com/Automattic/wp-calypso/pull/81816#discussion_r1336076011 and https://github.com/Automattic/wp-calypso/pull/81816#discussion_r1336073307.

If there are concerns, I'm more than happy to create a separate PR

sirbrillig replied:
Looks fine to me. Thank you for implementing the suggestions!

sirbrillig commented on packages/shopping-cart/src/types.ts:
```diff
@@ -614,6 +614,7 @@ export interface RequestCartProductExtra extends ResponseCartProductExtra {
 	import_dns_records?: boolean;
 	signup?: boolean;
 	headstart_theme?: string;
+	feature_slug?: string;
```
Your comment above made me think that it might be a good idea to add a comment directly above this property in the type to explain its purpose. That way the next developer to see it will know what it's for without having to search for this PR. (Honestly we should do that for every property here but we've got to start somewhere.)

southp reviewed (approved):
Thanks for the great collaboration on making the code even better, @jeyip @sirbrillig . 

I've confirmed the both new events work as expected 🚢

southp commented:
@jeyip FYI we want to expedite the launch so I'm going ahead and deploying this :)
